### ISSUE TITLE
Smarter computation of plugin dependencies.

### ISSUE THREAD

**Issue by ditman** [flutter/flutter#142276]
### Use case

In the current implementation of our plugin-related tooling, all compatible plugin dependencies of a plugin are added to the generated plugin registrar of the platform, regardless of the plugin itself actually supports the target platform.

A concrete example:

if I add `firebase_crashlytics` to my pubspec (this is only compatible Android/iOS/MacOS) and compile for the web, our build system will add `firebase_core` to the plugin web registrant (because it *is* web compatible), but will never add `crashlytics` (which is *not* web compatible), hence making `firebase_core` unnecessary.

The web plugin registrant generates Dart code that will always call the `registerWith` method of the unused plugin, so at least parts of it will fail to tree-shake, increasing the binary size.

(Another problem on the web is that this impedes compiling to `--wasm`, **even if** the App code itself has been written correctly (conditional imports, etc...) to never call the plugin code paths when targeting the web)

### Proposal

Modify the generation of the `plugin_registrar` to **only include dependencies of the plugins that will be `registerWith` for that platform**, and don't include all its plugin dependencies blindly.

**Comment by ditman** [comment]
I verified that this happens on the web, is this a thing in all platforms? I don't mind fixing this for `platform-web` only.

**Comment by ditman** [comment]
(I don't think that a plugin depending on another plugin is a very common occurrence, so this is possibly a micro-optimization, but the Wasm compilation breaking has just happened to me attempting to compile flutter/cocoon!)

**Comment by ditman** [comment]
TL;DR: This may need to be modified: 

https://github.com/flutter/flutter/blob/15fa68ab1dea1a5c20419e611f7881b32b88e9e1/packages/flutter_tools/lib/src/flutter_plugins.dart#L80-L105

**Comment by ditman** [comment]
I verified that this happens on the web output for sure, and looking at the code above, it's probably "a thing" in all platforms.

I don't think that a plugin depending on another plugin is a very common occurrence, so this is possibly a micro-optimization. (But the Wasm _problem_ just bit me attempting to compile flutter/cocoon.)

**Comment by stuartmorgan** [comment]
(See also https://github.com/flutter/flutter/issues/81650)

**Comment by ditman** [comment]
Ah yes, I wasn't trying to minimize the complexity of this problem, just trying to describe what I was seeing in this particular case.

Whomever ends up implementing this, listen to Stuart 😜

(PS, in the case of the "undendorsed version of A (uA)", if uA uses B because it's available because of A, isn't that a (bad) transient dependency, and uA should be fixed to depend on B?)

**Comment by stuartmorgan** [comment]
> (PS, in the case of the "undendorsed version of A (uA)", if uA uses B because it's available because of A, isn't that a (bad) transient dependency, and uA should be fixed to depend on B?)

I wasn't think of uA using B, but of app-facing-A using B. If we trim B under the theory that if A doesn't support platform P, clients won't be calling A (and thus won't need B) on P, that would be wrong once uA (supporting P) is added to the mix, because now the client *will* be calling A on P.

**Comment by ditman** [comment]
Ah, you're right, adding platforms through unendorsed packages should be taken into account when culling the deps too!
